### ISSUE TITLE
[PROPOSAL] post-restructuring PROPOSAL output confuses I3MuonSlicer (Trac #2340)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2332]
I3MuonSlicer was designed to reverse-engineer the MMC output format to recover the energy of the muon between stochastic losses. When run against new PROPOSAL simulation, I3MuonSlicer inserts a muon segment ''after'' a muon decays, like so:

```text
    23 MuPlus (1117.15m, 60.9397m, 1949.98m) (30.0081deg, 359.703deg) 11.5111ns 429.205GeV 1546.08m
      73 MuPlus (453.863m, 64.374m, 801.499m) (30.0081deg, 359.703deg) 4435.44ns 50.5399GeV 28.1962m
      15 DeltaE (439.762m, 64.447m, 777.083m) (30.0081deg, 359.703deg) 4523.72ns 1.73634GeV 0m
      74 MuPlus (439.762m, 64.447m, 777.083m) (30.0081deg, 359.703deg) 4529.49ns 46.994GeV 76.9354m
      16 DeltaE (401.285m, 64.6462m, 710.46m) (30.0081deg, 359.703deg) 4780.35ns 0.625515GeV 0m
      75 MuPlus (401.285m, 64.6462m, 710.46m) (30.0081deg, 359.703deg) 4786.12ns 41.4309GeV 114.689m
      17 EPlus (343.927m, 64.9432m, 611.144m) (180deg, 180deg) 6928.48ns 0.0504133GeV 0m
      18 NuE (343.927m, 64.9432m, 611.144m) (180deg, 180deg) 6928.48ns 0.0393494GeV 0m
      19 NuMuBar (343.927m, 64.9432m, 611.144m) (180deg, 180deg) 6928.48ns 0.0158957GeV 0m
      78 MuPlus (343.927m, 64.9432m, 611.144m) (30.0081deg, 359.703deg) 5168.69ns 33.9646GeV 527.572m
```

This really screws up stopping muons, since instead of stopping, they burst back to life. A likely culprit is MMCTrack. For MMC (and PROPOSAL Classic), stopping tracks had negative values for Ef, which were interpreted as -1 times the distance from the decay to trajectory exit point in cm (hooray, units!). For example

```text
[I3MMCTrack = [
 (xi, yi, zi, ti, Ei) = (404.223 ,-0.996252 ,800 ,4522.09 ,0.338673)
 (xc, yc, zc, tc, Ec) = (-44.3509 ,-0.849819 ,25.6933 ,5404.34 ,0)
 (xf, yf, zf, tf, Ef) = (-522.694 ,-0.693668 ,-800 ,5404.34 ,-1330.06)
 Elost = 0.338673
 Particle = [ I3Particle MajorID : 2817278544529359988
             MinorID : 22
              Zenith : 0.525077
             Azimuth : 6.28286
                   X : 1070.44
                   Y : -1.21373
                   Z : 1949.99
                Time : 88.7298
              Energy : 407.132
               Speed : 0.299792
              Length : 1330.06
                Type : MuPlus
        PDG encoding : -13
               Shape : StartingTrack
              Status : NotSet
            Location : InIce
]]
```

For reasons that probably seemed reasonable at the time, new PROPOSAL sets the checkpoint energies to the muon rest mass:

```text
[I3MMCTrack = [
 (xi, yi, zi, ti, Ei) = (452.997 ,64.3785 ,800 ,4435.44 ,50.5399)
 (xc, yc, zc, tc, Ec) = (343.927 ,64.9432 ,611.144 ,6928.48 ,0.105658)
 (xf, yf, zf, tf, Ef) = (343.927 ,64.9432 ,611.144 ,6928.48 ,0.105658)
 Elost = 50.4343
 Particle = [ I3Particle MajorID : 2817278544529359988
             MinorID : 23
              Zenith : 0.52374
             Azimuth : 6.27801
                   X : 1117.15
                   Y : 60.9397
                   Z : 1949.98
                Time : 11.5111
              Energy : 429.205
               Speed : 0.299792
              Length : 1546.08
                Type : MuPlus
        PDG encoding : -13
               Shape : StartingTrack
              Status : NotSet
            Location : InIce
]]
```

Since there are multiple places in IceSim where I3MMCTracks are used to infer the energies of muons, and no general way to detect which edition of PROPOSAL created a given I3MMCTrack, it would best to make the new output values consistent with MMC.

Migrated from https://code.icecube.wisc.edu/ticket/2340
```json
{
    "status": "closed", 
    "changetime": "2019-09-16T09:08:57", 
    "description": "I3MuonSlicer was designed to reverse-engineer the MMC output format to recover the energy of the muon between stochastic losses. When run against new PROPOSAL simulation, I3MuonSlicer inserts a muon segment ''after'' a muon decays, like so:\n\n{{{\n    23 MuPlus (1117.15m, 60.9397m, 1949.98m) (30.0081deg, 359.703deg) 11.5111ns 429.205GeV 1546.08m\n      73 MuPlus (453.863m, 64.374m, 801.499m) (30.0081deg, 359.703deg) 4435.44ns 50.5399GeV 28.1962m\n      15 DeltaE (439.762m, 64.447m, 777.083m) (30.0081deg, 359.703deg) 4523.72ns 1.73634GeV 0m\n      74 MuPlus (439.762m, 64.447m, 777.083m) (30.0081deg, 359.703deg) 4529.49ns 46.994GeV 76.9354m\n      16 DeltaE (401.285m, 64.6462m, 710.46m) (30.0081deg, 359.703deg) 4780.35ns 0.625515GeV 0m\n      75 MuPlus (401.285m, 64.6462m, 710.46m) (30.0081deg, 359.703deg) 4786.12ns 41.4309GeV 114.689m\n      17 EPlus (343.927m, 64.9432m, 611.144m) (180deg, 180deg) 6928.48ns 0.0504133GeV 0m\n      18 NuE (343.927m, 64.9432m, 611.144m) (180deg, 180deg) 6928.48ns 0.0393494GeV 0m\n      19 NuMuBar (343.927m, 64.9432m, 611.144m) (180deg, 180deg) 6928.48ns 0.0158957GeV 0m\n      78 MuPlus (343.927m, 64.9432m, 611.144m) (30.0081deg, 359.703deg) 5168.69ns 33.9646GeV 527.572m\n}}}\n\nThis really screws up stopping muons, since instead of stopping, they burst back to life. A likely culprit is MMCTrack. For MMC (and PROPOSAL Classic), stopping tracks had negative values for Ef, which were interpreted as -1 times the distance from the decay to trajectory exit point in cm (hooray, units!). For example\n\n{{{\n[I3MMCTrack = [\n (xi, yi, zi, ti, Ei) = (404.223 ,-0.996252 ,800 ,4522.09 ,0.338673)\n (xc, yc, zc, tc, Ec) = (-44.3509 ,-0.849819 ,25.6933 ,5404.34 ,0)\n (xf, yf, zf, tf, Ef) = (-522.694 ,-0.693668 ,-800 ,5404.34 ,-1330.06)\n Elost = 0.338673\n Particle = [ I3Particle MajorID : 2817278544529359988\n             MinorID : 22\n              Zenith : 0.525077\n             Azimuth : 6.28286\n                   X : 1070.44\n                   Y : -1.21373\n                   Z : 1949.99\n                Time : 88.7298\n              Energy : 407.132\n               Speed : 0.299792\n              Length : 1330.06\n                Type : MuPlus\n        PDG encoding : -13\n               Shape : StartingTrack\n              Status : NotSet\n            Location : InIce\n]]\n}}}\n\nFor reasons that probably seemed reasonable at the time, new PROPOSAL sets the checkpoint energies to the muon rest mass:\n\n{{{\n[I3MMCTrack = [\n (xi, yi, zi, ti, Ei) = (452.997 ,64.3785 ,800 ,4435.44 ,50.5399)\n (xc, yc, zc, tc, Ec) = (343.927 ,64.9432 ,611.144 ,6928.48 ,0.105658)\n (xf, yf, zf, tf, Ef) = (343.927 ,64.9432 ,611.144 ,6928.48 ,0.105658)\n Elost = 50.4343\n Particle = [ I3Particle MajorID : 2817278544529359988\n             MinorID : 23\n              Zenith : 0.52374\n             Azimuth : 6.27801\n                   X : 1117.15\n                   Y : 60.9397\n                   Z : 1949.98\n                Time : 11.5111\n              Energy : 429.205\n               Speed : 0.299792\n              Length : 1546.08\n                Type : MuPlus\n        PDG encoding : -13\n               Shape : StartingTrack\n              Status : NotSet\n            Location : InIce\n]]\n}}}\n\nSince there are multiple places in IceSim where I3MMCTracks are used to infer the energies of muons, and no general way to detect which edition of PROPOSAL created a given I3MMCTrack, it would best to make the new output values consistent with MMC.", 
    "reporter": "jvansanten", 
    "cc": "kjmeagher", 
    "resolution": "fixed", 
    "_ts": "1568624937075517", 
    "component": "combo simulation", 
    "summary": "[PROPOSAL] post-restructuring PROPOSAL output confuses I3MuonSlicer", 
    "priority": "blocker", 
    "keywords": "", 
    "time": "2019-08-07T15:57:56", 
    "milestone": "Autumnal Equinox 2019", 
    "owner": "jsoedingrekso", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20190807T16:18:45: 
* olivas changed priority from "major" to "blocker"
* olivas changed cc from "kjmeagher, jsoedingrekso" to "kjmeagher"
* olivas changed status from "new" to "assigned"
* olivas changed owner from "" to "jsoedingrekso"

**Comment by icecube-trac** [comment]
Trac update at 20190807T16:22:34: 
* jvansanten changed _comment0 from "For reference, here is the relevant bit of the Amanda that special-cased Ef for stopping tracks: https://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/PROPOSAL/releases/V00-00-02/private/PROPOSAL/Amanda.cxx#L2364" to "1565194979868508"
* jvansanten commented:

> For reference, here is the relevant bit of MMC/PROPOSAL Classic's Amanda that special-cased Ef for stopping tracks: https://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/PROPOSAL/releases/V00-00-02/private/PROPOSAL/Amanda.cxx#L2364

**Comment by icecube-trac** [comment]
Trac update at 20190808T05:41:09: jsoedingrekso changed status from "assigned" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20190808T06:47:12: jsoedingrekso commented:

> In my opinium this feature is absolutely counterintuitive and I don't want to force other experiments using PROPOSAL to adapt to this artefact.
> It's more a question, if we really need the MuonSlicer anymore, since it's not even working for cascades placed 1mm away from the initial track direction (which could happen due to MultipleScattering and what is the reason I turn off MultipleScattering in 174782).
> There is actually the option for a continuous loss output, which could replace the MuonSlicer, but I haven't looked up how exactly the muon segments are created eg. what time the muon segments will get. It should be the same as the time of the previous secondary, but in the example above it isn't. Maybe because the time in the MuonSlicer is just calculated with dx/c and not with the time integral (considering the decreasing energy and travel speed). But this is just guessing; haven't looked at this in more detail, yet.

**Comment by icecube-trac** [comment]
Trac update at 20190808T08:43:18: jvansanten commented:

> I agree completely that the feature is counterintuitive, and am not suggesting that it creep out into PROPOSAL proper. `I3PropagatorServicePROPOSAL` is however the IceCube-specific interface to PROPOSAL, and when you emit an `I3MMCTrack`, you enter an implicit contract to mimic the output of MMC. If we want to change the meaning of the fields, we should introduce a new type so that there can be no confusion.
> 
> In the longer term I would also like to eliminate `I3MuonSlicer`, for exactly the reasons you mentioned. Trunk clsim in fact consumes secondaries and continuous segments directly from PROPOSAL when requested, but I am reluctant to break compatibility with older generation strategies where muon propagation is grouped with event generation. Making `I3PropagatorServicePROPOSAL` fill `I3MMCTrack` with values in the MMC convention seems like the easiest way to get this functionality back.
> 
> In addition, `I3MuonGun::Track` uses `I3MMCTrack` to provide estimates of the muon energy at any point along its trajectory. This is also somewhat tricky, as the history of bug-fixes indicates, but is extraordinarily useful to have. I would be happy to see `I3MMCTrack` eliminated if we could come up with a data structure that summarizes a track's history at a level of detail similar to the current `I3MCTree` + `I3MMCTrack`, ideally while being much smaller.

**Comment by icecube-trac** [comment]
Trac update at 20190814T18:06:47: 
* jsoedingrekso commented:

> In changeset 175045:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="175045"
> change PROPOSAL output/MMCTrack to old behavior and set Ef of a decaying particle to negativ distance till the detector border. Should fix #2332
> ```


* jsoedingrekso changed status from "accepted" to "closed"
* jsoedingrekso changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190916T09:08:57: jvansanten commented:

> Following on to this with further conventions from MMC:
> - If the particle decays inside the active volume, [https://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/PROPOSAL/releases/V00-00-02/private/PROPOSAL/Amanda.cxx#L2391 Ec is 0] (which makes more sense than the rest mass, because the particle does not exist in the final state)
> - The point of closest approach (xc,yc,zc,tc) is calculated from a [https://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/PROPOSAL/releases/V00-00-02/private/PROPOSAL/Amanda.cxx#L1999 secret dot product between the displacement and track direction], and is set [https://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/PROPOSAL/releases/V00-00-02/private/PROPOSAL/Amanda.cxx#L2369 whether or not the particle ever got there]
